test/cli_test.go

```diff
@@ -0,0 +1,379 @@
+
+			terraform(t, dir, c.name)
+
+			t.Run("cmd=convert", func(t *testing.T) {
+				testConvertCommand(t, dir, c.name, c.offline)
+			})
+
+			for _, r := range c.rules {
+				t.Run(fmt.Sprintf("cmd=validate/rule=%s", r.name), func(t *testing.T) {
+					testValidateCommand(t, r.wantError, r.wantOutputRegex, dir, c.name, c.offline, r.name)
+				})
+			}
+		})
+	}
+}
+
+func testConvertCommand(t *testing.T, dir, name string, offline bool) {
+	var payload []byte
+	switch version.LeastSupportedVersion() {
+	case version.TF11:
+		payload = tfvConvert(t, dir, name+".tfplan", offline)
+	default:
+		payload = tfvConvert(t, dir, name+".tfplan.json", offline)
+	}
+	// Verify if the generated assets match the expected.
+	var got []google.Asset
+	err := json.Unmarshal(payload, &got)
+	if err != nil {
+		t.Fatalf("unmarshaling: %v", err)
+	}
+	testfile := filepath.Join(dir, name+".json")
+	payload, err = ioutil.ReadFile(testfile)
+	if err != nil {
+		t.Fatalf("Error reading %v: %v", testfile, err)
+	}
+	var want []google.Asset
+	if err := json.Unmarshal(payload, &want); err != nil {
+		t.Fatalf("unmarshaling: %v", err)
+	}
+
+	if !offline {
+		// remove the ancestry as the value of that is dependent on project,
+		// and is not important for the test.
+		for i := range got {
+			got[i].Ancestry = ""
+		}
+		for i := range want {
+			want[i].Ancestry = ""
+		}
+	}
+	// compare assets
+	gotJSON, err := json.Marshal(got)
+	if err != nil {
+		t.Fatalf("marshaling: %v", err)
+	}
+	wantJSON, err := json.Marshal(want)
+	if err != nil {
+		t.Fatalf("marshaling: %v", err)
+	}
+	require.JSONEq(t, string(wantJSON), string(gotJSON))
+}
+
+func testValidateCommand(t *testing.T, wantError bool, want, dir, name string, offline bool, ruleName string) {
+	cwd, err := os.Getwd()
+	if err != nil {
+		t.Fatalf("cannot get current directory: %v", err)
+	}
+	policyPath := filepath.Join(cwd, samplePolicyPath, ruleName)
+	var got []byte
+	switch version.LeastSupportedVersion() {
+	case version.TF11:
+		got = tfvValidate(t, wantError, dir, name+".tfplan", policyPath, offline)
+	default:
+		got = tfvValidate(t, wantError, dir, name+".tfplan.json", policyPath, offline)
+	}
+	wantRe := regexp.MustCompile(want)
+	if want != "" && !wantRe.Match(got) {
+		t.Fatalf("binary did not return expect output, \ngot=%s \nwant (regex)=%s", string(got), want)
+	}
+}
+
+func terraform(t *testing.T, dir, name string) {
+	switch version.LeastSupportedVersion() {
+	case version.TF11:
+		terraformInit(t, "terraform", dir)
+		terraformPlan(t, "terraform", dir, name+".tfplan")
+	default:
+		terraformInit(t, "terraform", dir)
+		terraformPlan(t, "terraform", dir, name+".tfplan")
+		payload := terraformShow(t, "terraform", dir, name+".tfplan")
+		saveFile(t, dir, name+".tfplan.json", payload)
+	}
+}
+
+func terraformInit(t *testing.T, executable, dir string) {
+	wantError := false
+	cmd := exec.Command(executable, "init", "-input=false", dir)
+	cmd.Env = []string{"HOME=" + filepath.Join(dir, "fakehome")}
```

Comment:
There seems to be a lot of repeated code between `terraformInit`, `terraformPlan`, etc.

Could we consolidate with a `terraformExec` function and simply call that multiple times with different commands?

Reply:
yes. Fixed.